File: leetcode/design/1381. Design a Stack With Increment Operation/py/main.py
class CustomStack:
    # Time O(1)
    # Space O(maxSize)
    def __init__(self, maxSize: int):
        self.st = []
        self.maxSize = maxSize
        # Use an additional array to record the increment value.
        self.inc = []

    def push(self, x: int) -> None:
        if len(self.st) < self.maxSize:
            self.st.append(x)
            self.inc.append(0)

    def pop(self) -> int:
        if self.isEmpty():
            return -1

        if len(self.inc) > 1:
            self.inc[-2] += self.inc[-1]

        return self.st.pop() + self.inc.pop()
    
    def isEmpty(self):
        return len(self.inc) == 0

    def increment(self, k: int, val: int) -> None:
        if self.inc:
            self.inc[min(k, len(self.inc)) - 1] += val 

File: leetcode/design/1381. Design a Stack With Increment Operation/js/main.js
class CustomStack {
  constructor(maxSize) {
    this.stack = [];
    this.maxSize = maxSize;
  }

  push(x) {
    if (this.stack.length < this.maxSize) {
      this.stack.push(x);
    }
  }

  pop() {
    if (this.stack.length == 0) return -1;
    return this.stack.pop();
  }

  increment(k, val) {
    let i = 0;

    while (i < this.stack.length && k > 0) {
      this.stack[i] = this.stack[i] + val;
      k--;
      i++;
    }
  }
}
